File: backend/src/controllers/admin.controller.ts
// backend/src/controllers/admin.controller.ts
import type { Request, Response, NextFunction } from 'express';
import * as svc from '../services/admin.service';

const wrap =
  (fn: (req: Request, res: Response) => Promise<unknown>) =>
  async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try { await fn(req, res); } catch (err) { next(err); }
  };

export const getPlatformStats = wrap(async (_req, res) => {
  res.json(await svc.getPlatformStats());
});

export const getAnalyticsOverview = wrap(async (_req, res) => {
  res.json(await svc.getAnalyticsOverview());
});
export const getAnalyticsLoans = wrap(async (_req, res) => {
  res.json(await svc.getAnalyticsLoans());
});
export const getAnalyticsAssets = wrap(async (_req, res) => {
  res.json(await svc.getAnalyticsAssets());
});
export const getAnalyticsGrowth = wrap(async (_req, res) => {
  res.json(await svc.getAnalyticsGrowth());
});

export const listClubs = wrap(async (req, res) => {
  const page   = Math.max(1, Number(req.query.page)  || 1);
  const limit  = Math.min(100, Math.max(1, Number(req.query.limit) || 20));
  const search = typeof req.query.search === 'string' ? req.query.search : undefined;
  res.json(await svc.listClubs(page, limit, search));
});

export const getClubDetail = wrap(async (req, res) => {
  res.json(await svc.getClubDetail(req.params.id));
});

export const updateClubStatus = wrap(async (req, res) => {
  const { is_active } = req.body;
  if (typeof is_active !== 'boolean') {
    res.status(400).json({ statusCode: 400, error: 'Bad Request', message: 'is_active must be a boolean' });
    return;
  }
  await svc.updateClubStatus(req.params.id, is_active);
  res.json({ message: 'Club status updated' });
});

export const resetClubAdminPassword = wrap(async (req, res) => {
  const temp_password = await svc.resetClubAdminPassword(req.params.id);
  res.json({ temp_password });
});

export const listClubUsers = wrap(async (req, res) => {
  const page  = Math.max(1, Number(req.query.page)  || 1);
  const limit = Math.min(100, Math.max(1, Number(req.query.limit) || 20));
  res.json(await svc.listClubUsers(req.params.id, page, limit));
});

export const updateUserStatus = wrap(async (req, res) => {
  const { is_active } = req.body;
  if (typeof is_active !== 'boolean') {
    res.status(400).json({ statusCode: 400, error: 'Bad Request', message: 'is_active must be a boolean' });
    return;
  }
  await svc.updateUserStatus(req.params.id, req.params.uid, is_active);
  res.json({ message: 'User status updated' });
});

export const resetUserPassword = wrap(async (req, res) => {
  const temp_password = await svc.resetUserPassword(req.params.id, req.params.uid);
  res.json({ temp_password });
});

export const listClubAssets = wrap(async (req, res) => {
  const page  = Math.max(1, Number(req.query.page)  || 1);
  const limit = Math.min(100, Math.max(1, Number(req.query.limit) || 20));
  res.json(await svc.listClubAssets(req.params.id, page, limit));
});

export const retireAsset = wrap(async (req, res) => {
  const { status } = req.body as { status?: string };
  if (status !== 'retired') {
    res.status(400).json({ statusCode: 400, error: 'Bad Request', message: 'status must be "retired"' });
    return;
  }
  await svc.retireAsset(req.params.id, req.params.aid);
  res.json({ message: 'Asset retired' });
});

export const deleteAsset = wrap(async (req, res) => {
  await svc.deleteAsset(req.params.id, req.params.aid);
  res.json({ message: 'Asset deleted' });
});

export const listClubLoans = wrap(async (req, res) => {
  const page   = Math.max(1, Number(req.query.page)  || 1);
  const limit  = Math.min(100, Math.max(1, Number(req.query.limit) || 20));
  const status = typeof req.query.status === 'string' ? req.query.status : undefined;
  res.json(await svc.listClubLoans(req.params.id, page, limit, status));
});

File: api/src/SportStock.Api/Controllers/AdminController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using SportStock.Api.Auth;
using SportStock.Api.Data.Enums;
using SportStock.Api.Dtos.Admin;
using SportStock.Api.Dtos.AuditLog;
using SportStock.Api.Dtos.SportType;
using SportStock.Api.Exceptions;
using SportStock.Api.Services;

namespace SportStock.Api.Controllers;

[ApiController]
[Authorize]
[Route("api/v1/admin")]
[RequireRole(UserRole.SuperAdmin)]
public sealed class AdminController(IAdminService service, IAuditLogService auditLogService, ISportTypeService sportTypeService) : ControllerBase
{
    [HttpGet("stats")]
    public async Task<IActionResult> Stats(CancellationToken ct) =>
        Ok(await service.GetPlatformStatsAsync(ct));

    [HttpGet("analytics/overview")]
    public async Task<IActionResult> AnalyticsOverview(
        [FromQuery] AnalyticsClubFilterQuery query, CancellationToken ct) =>
        Ok(await service.GetAnalyticsOverviewAsync(query.ClubId, ct));

    [HttpGet("analytics/loans")]
    public async Task<IActionResult> AnalyticsLoans(
        [FromQuery] AnalyticsClubFilterQuery query, CancellationToken ct) =>
        Ok(await service.GetAnalyticsLoansAsync(query.ClubId, ct));

    [HttpGet("analytics/assets")]
    public async Task<IActionResult> AnalyticsAssets(
        [FromQuery] AnalyticsClubFilterQuery query, CancellationToken ct) =>
        Ok(await service.GetAnalyticsAssetsAsync(query.ClubId, ct));

    [HttpGet("analytics/growth")]
    public async Task<IActionResult> AnalyticsGrowth(CancellationToken ct) =>
        Ok(await service.GetAnalyticsGrowthAsync(ct));

    [HttpGet("clubs")]
    public async Task<IActionResult> ListClubs(
        [FromQuery] ListClubsQuery query, CancellationToken ct) =>
        Ok(await service.ListClubsAsync(query, ct));

    [HttpGet("clubs/{id:guid}")]
    public async Task<IActionResult> GetClub(Guid id, CancellationToken ct) =>
        Ok(await service.GetClubAsync(id, ct));

    [HttpPatch("clubs/{id:guid}/status")]
    public async Task<IActionResult> UpdateClubStatus(
        Guid id, [FromBody] UpdateActiveRequest body, CancellationToken ct)
    {
        if (body.IsActive is null) throw new AppException("is_active must be a boolean", 400);
        await service.UpdateClubStatusAsync(id, body.IsActive.Value, ct);
        return Ok(new StatusMessageResponse { Message = "Club status updated" });
    }

    [HttpPost("clubs/{id:guid}/reset-admin-password")]
    public async Task<IActionResult> ResetClubAdmin(Guid id, CancellationToken ct)
    {
        var temp = await service.ResetClubAdminPasswordAsync(id, ct);
        return Ok(new TempPasswordResponse { TempPassword = temp });
    }

    [HttpGet("clubs/{id:guid}/users")]
    public async Task<IActionResult> ListClubUsers(
        Guid id, [FromQuery] int page, [FromQuery] int limit, CancellationToken ct) =>
        Ok(await service.ListClubUsersAsync(id, page == 0 ? 1 : page, limit == 0 ? 20 : limit, ct));

    [HttpPatch("clubs/{id:guid}/users/{uid:guid}/status")]
    public async Task<IActionResult> UpdateUserStatus(
        Guid id, Guid uid, [FromBody] UpdateActiveRequest body, CancellationToken ct)
    {
        if (body.IsActive is null) throw new AppException("is_active must be a boolean", 400);
        await service.UpdateUserStatusAsync(id, uid, body.IsActive.Value, ct);
        return Ok(new StatusMessageResponse { Message = "User status updated" });
    }

    [HttpPost("clubs/{id:guid}/users/{uid:guid}/reset-password")]
    public async Task<IActionResult> ResetUserPassword(Guid id, Guid uid, CancellationToken ct)
    {
        var temp = await service.ResetUserPasswordAsync(id, uid, ct);
        return Ok(new TempPasswordResponse { TempPassword = temp });
    }

    [HttpGet("clubs/{id:guid}/assets")]
    public async Task<IActionResult> ListClubAssets(
        Guid id, [FromQuery] ListClubResourcesQuery query, CancellationToken ct) =>
        Ok(await service.ListClubAssetsAsync(id, query, ct));

    [HttpPatch("clubs/{id:guid}/assets/{aid:guid}/status")]
    public async Task<IActionResult> UpdateAssetStatus(
        Guid id, Guid aid, [FromBody] UpdateActiveRequest body, CancellationToken ct)
    {
        if (body.IsActive is null) throw new AppException("is_active must be a boolean", 400);
        await service.UpdateAssetStatusAsync(id, aid, body.IsActive.Value, ct);
        var msg = body.IsActive.Value ? "Asset enabled" : "Asset disabled";
        return Ok(new StatusMessageResponse { Message = msg });
    }

    [HttpDelete("clubs/{id:guid}/assets/{aid:guid}")]
    public async Task<IActionResult> DeleteAsset(Guid id, Guid aid, CancellationToken ct)
    {
        await service.DeleteAssetAsync(id, aid, ct);
        return Ok(new StatusMessageResponse { Message = "Asset deleted" });
    }

    [HttpGet("clubs/{id:guid}/loans")]
    public async Task<IActionResult> ListClubLoans(
        Guid id,
        [FromQuery] int page,
        [FromQuery] int limit,
        [FromQuery] string? status,
        CancellationToken ct) =>
        Ok(await service.ListClubLoansAsync(id, page == 0 ? 1 : page, limit == 0 ? 20 : limit, status, ct));

    // GET /api/v1/admin/audit-logs — super-admin view across all clubs
    [HttpGet("audit-logs")]
    public async Task<IActionResult> ListAll([FromQuery] AuditLogQuery q)
        => Ok(await auditLogService.ListAsync(q, clubId: null));

    // ── Sport-type settings ──────────────────────────────────────────────────

    [HttpGet("settings/sport-types")]
    public async Task<IActionResult> AdminListSportTypes()
        => Ok(await sportTypeService.ListAllAsync());

    [HttpPost("settings/sport-types")]
    public async Task<IActionResult> CreateSportType([FromBody] CreateSportTypeRequest req)
        => StatusCode(201, await sportTypeService.CreateAsync(req));

    [HttpPut("settings/sport-types/{id:guid}")]
    public async Task<IActionResult> UpdateSportType(Guid id, [FromBody] UpdateSportTypeRequest req)
    {
        await sportTypeService.UpdateAsync(id, req);
        return NoContent();
    }

    [HttpDelete("settings/sport-types/{id:guid}")]
    public async Task<IActionResult> DeleteSportType(Guid id)
    {
        await sportTypeService.DeleteAsync(id);
        return NoContent();
    }
}
